src/doc/trpl/dining-philosophers.md

```diff
@@ -0,0 +1,942 @@
+output:
+
+```text
+Baruch Spinoza is done eating.
+Gilles Deleuze is done eating.
+Karl Marx is done eating.
+Friedrich Nietzsche is done eating.
+Michel Foucault is done eating.
+```
+
+Easy enough, they’re all done! We haven’t actually implemented the real problem
+yet, though, so we’re not done yet!
+
+The next version is a bit larger. Here it is:
+
+```rust
+use std::sync::mpsc;
+use std::sync::mpsc::Sender;
+
+struct Philosopher {
+    name: String,
+    done: Sender<bool>
+}   
+
+impl Philosopher { 
+    fn new(name: &str, done: Sender<bool>) -> Philosopher {
+        Philosopher {
+            name: name.to_string(),
+            done: done,
+        }
+    }
+    
+    fn done(&self) {
+        println!("{} is done eating.", self.name);
+
+        self.done.send(true).ok().expect("Couldn't finish eating");
+    }
+}
+
+fn main() {
+    let (done_tx, done_rx) = mpsc::channel();
+
+    let philosophers = vec![
+        Philosopher::new("Baruch Spinoza", done_tx.clone()),
+        Philosopher::new("Gilles Deleuze", done_tx.clone()),
+        Philosopher::new("Karl Marx", done_tx.clone()),
+        Philosopher::new("Friedrich Nietzsche", done_tx.clone()),
+        Philosopher::new("Michel Foucault", done_tx.clone()),
+    ];
+
+    for p in &philosophers {
+        p.done();
+    }
+
+    for _ in 0..5 {
+        done_rx.recv().unwrap();
+    }
+}
+```
+
+Whew! Lots there. Let’s go through it:
+
+```rust,ignore
+use std::sync::mpsc;
+use std::sync::mpsc::Sender;
+```
+
+The `use` keyword lets us bring names into scope. In this case, we are going to
+use two things: the `mspc` module from `std::sync`, and the `Sender` type from
+said module. As you may have guessed, `mspc` stands for something:
+multi-producer, single consumer. We’re going to be using Rust’s ‘channel’ types
+to handle the concurrency here, and they’re in this part of the standard
+library.
+
+```rust,ignore
+struct Philosopher {
+    name: String,
+    done: Sender<bool>
+}   
+```
+
+When our philosopher is done eating, they’re going to send a boolean value
+down a channel. The sending end of a channel has the `Sender` type, which
+we `use`d in the last line.
+
+```rust,ignore
+impl Philosopher { 
+    fn new(name: &str, done: Sender<bool>) -> Philosopher {
+        Philosopher {
+            name: name.to_string(),
+            done: done,
+        }
+    }
+```
+
+Now that we have a second field for our struct, we need to define it in
+`new()`. In this case, we’re going to take it as an argument. We’ll pass this
+in in `main()`, below.
+    
+```rust,ignore
+    fn done(&self) {
+        println!("{} is done eating.", self.name);
+
+        self.done.send(true).ok().expect("Couldn't finish eating");
+    }
+}
+```
+
+We’ve added a new line to `done()`. We use the `send()` method on `Sender`
+to send `true` down the channel. `send()` returns a `Result`, letting us
+know if we’ve had some sort of error while trying to send the value. In
```

Comment:
I think it may be worth elaborating on what "some sort of error" is in this case as there's precisely one possible error, which is that the receiving end has disconnected (been dropped). In other words you will receive `Err(T)` (where you sent the `T`) if it is 100% guaranteed that it will never be received.